rtl/eclrtl/rtlnewkey.hpp

```diff
@@ -0,0 +1,76 @@
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef RTLNEWKEY_INCL
+#define RTLNEwKEY_INCL
+#include "eclrtl.hpp"
+
+enum TransitionMask : byte
+{
+    CMPlt = 0x01,
+    CMPeq = 0x02,
+    CMPgt = 0x04,
+    CMPmin = 0x08,      //minimum possible value
+    CMPmax = 0x10,      //maximum possible value
+    CMPle = CMPlt | CMPeq,
+    CMPge = CMPgt | CMPeq,
+    CMPminmask = CMPgt|CMPmin,
+    CMPmaxmask = CMPlt|CMPmax,
+};
+inline constexpr TransitionMask operator | (TransitionMask l, TransitionMask r) { return (TransitionMask)((unsigned)l | (unsigned)r); } \
+
+class ValueTransition;
+interface RtlTypeInfo;
+interface IValueSet : public IInterface
+{
+//The following methods are used for creating a valueset
+    virtual ValueTransition * createTransition(TransitionMask mask, unsigned __int64 value) const = 0;
+    virtual ValueTransition * createStringTransition(TransitionMask mask, size32_t len, const char * value) const = 0;
+    virtual ValueTransition * createUtf8Transition(TransitionMask mask, size32_t len, const char * value) const = 0;
+    virtual void addRange(ValueTransition * loval, ValueTransition * hival) = 0;
+    virtual void addAll() = 0;
+    virtual void killRange(ValueTransition * loval, ValueTransition * hival) = 0;
+    virtual void reset() = 0;
+    virtual void invertSet() = 0;
+    virtual void unionSet(const IValueSet *) = 0;
+    virtual void excludeSet(const IValueSet *) = 0;
+    virtual void intersectSet(const IValueSet *) = 0;
+    virtual StringBuffer & serialize(StringBuffer & out) const= 0;
+
+//The following methods are for use once the value set has been created.
+    virtual unsigned numRanges() const = 0;
+
+    //find the last range where the lower bound <= the field, returns 0 if the field matches the lower bound, > 0 otherwise.
+    //matchRange is set to the range number, set to numRanges if there is no possible match.  Uses a binary chop
+    virtual int findCandidateRange(const byte * field, unsigned & matchRange) const = 0;
+
+    //find the last range where the lower bound <= the field, returns 0 if the field matches the lower bound, > 0 otherwise.
+    //starts searching from curRange (which is likely to match).  Uses a sequential search.
+    virtual int checkCandidateRange(const byte * field, unsigned & curRange) const = 0;
+
+    // Does this field match any range?
+    virtual bool matches(const byte * field) const = 0;
+
+    // Does this field match this particular range?
+    virtual bool matches(const byte * field, unsigned range) const = 0;
+
+    //Should this belong elsewhere?
```

Comment:
This comment seems to be targeting a different audience to the others...

Reply:
The issue being I think that this method is just a passthrough to the associated type compare function, so why not provide a way to retrieve that typeInfo structure instead?